.github/scripts/gitutils.py

```diff
@@ -331,3 +334,43 @@ def patterns_to_regex(allowed_patterns: List[str]) -> Any:
                 rc += c
     rc += ")"
     return re.compile(rc)
+
+# Modified from https://github.com/pytorch/pytorch/blob/b00206d4737d1f1e7a442c9f8a1cadccd272a386/torch/hub.py#L129
+def _read_url(url: Any) -> Any:
+    with urlopen(url) as r:
+        return r.headers, r.read().decode(r.headers.get_content_charset('utf-8'))
+
+
+def request_for_labels(url: str) -> Any:
+    headers = {'Accept': 'application/vnd.github.v3+json'}
+    return _read_url(Request(url, headers=headers))
+
+
+def get_last_page(header: Any) -> int:
+    # Link info looks like: <https://api.github.com/repositories/65600975/labels?per_page=100&page=2>;
+    # rel="next", <https://api.github.com/repositories/65600975/labels?per_page=100&page=3>; rel="last"
+    link_info = header['link']
+    prefix = "&page="
+    suffix = ">;"
+    return int(link_info[link_info.rindex(prefix) + len(prefix):link_info.rindex(suffix)])
+
+
+def update_labels(labels: List[str], info: str) -> None:
+    labels_json = json.loads(info)
+    labels.extend([x["name"] for x in labels_json])
+
+
+@lru_cache()
+def get_pytorch_labels() -> List[str]:
+    prefix = "https://api.github.com/repos/pytorch/pytorch/labels?per_page=100"
+    header, info = request_for_labels(prefix + "&page=1")
+    labels: List[str] = []
+    update_labels(labels, info)
+
+    last_page = get_last_page(header)
+    assert last_page > 0, "Error reading header info to determine total number of pages of labels"
+    for page_number in range(2, last_page + 1):  # skip page 1
+        _, info = request_for_labels(prefix + f"&page={page_number}")
+        update_labels(labels, info)
+
+    return labels
```

Comment:
Considering that trymerge replies heavily on GraphQL, wouldn't it be better to use GraphQL queries here as well:
```
query {
  repository(owner: "pytorch", name: "pytorch") {
    labels(first: 100) {
      nodes {
        name
      }
    }
  }
}
```

Reply:
Given this helper is in gitutils not trymerge and used in other scripts too, it's probably ok to leave it as is with the python implementation for now. I believe the underlying point is great that we want to make all the gh helper queries consistent across all scripts (atm it feels like each script is written on its own with some duplications and inconsistency), and ideally extract them out into a helper module, so all scripts can import and use them (instead of importing from each other). Do you think this would be worth having a separate issue to do a proper refactoring?